### DISCUSSION TITLE
2주차 : `mvc` 미션을 진행하며 생각해 볼까요?

### DISCUSSION THREAD

**Discussion by goldentrash** [gdsc-konkuk/24-25-study-spring-learning-with-test#19] [Assignments]
# 생각해 볼까요?

> 다음은 알아두면 재밌을, 어쩌면 몰라도 크게 상관없을 내용들이기도 합니다.
> 관심이 있는 주제(최소 1개)에 대해서 더 알아보고 스터디원과 생각을 교류해 봅시다!

### 1. 전통적인 MVC 구조를 알고 계신가요?

- Template Engine을 사용해 보신 경험이 있나요?

### 3. `Exception`과 `RuntimeException`의 차이에 대해 알고 계신가요?

- [Spring의 Exception Handling Process](https://mangkyu.tistory.com/204)에 대해 알고 계신가요?
- `ControllerAdvice`를 사용해보셨나요?

### 4. HTTP Status Code를 잘 활용하고 계신가요?

- 애용하는 Status Code가 있나요?
- 본인에게는 익숙하지만, 누군가에겐 조금 생소할 만한 Code가 있다면 공유해주시겠어요? (like `429` [Too Many Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429))
- [여러분은 몇 번 코드가 옳다고 생각하시나요?](https://techblog.yogiyo.co.kr/%EB%8D%B0%EC%9D%B4%ED%84%B0%EA%B0%80-%EC%97%86%EC%9D%84-%EB%95%8C-200%EC%9D%B8%EA%B0%80-404%EC%9D%B8%EA%B0%80-f1c8c39ca9df) (이건 정말 콜로세움 열립니다 ㅋㅋ)
   
### 6. TLS/SSL에 대해 알고 계신가요?

- HTTPs는 Request/Response의 어떤 부분을 암호화하나요? 전체? Body? 어쩌면 Start Line을 제외한 나머지?
    
### 7. OpenAPI spec에 대해 들어보셨나요?

- (최신) Swagger는 GET request의 payload가 request body에 담기는 경우를 허용하지 않는답니다...
- [JSON reponse spec에 대한 논의](https://stackoverflow.com/questions/12806386/is-there-any-standard-for-json-api-response-format)도 재답니다.

# 아래에 정리해 주세요!

> 정리 양식은 자유입니다! 몇 가지 주제에 대해서 알아오셨나요?

**Comment by sjin0911** [comment]
## 1. 전통적인 MVC 구조를 알고 계신가요?

- Template Engine을 사용해 보신 경험이 있나요?
    - Template Engine: 템플릿 양식과 특정 데이터 모델에 따른 입력 자료를 합성하여 결과 문서를 출력하는 소프트웨어 컴포넌트
    - Web Template Engine: 웹 문서가 출력되는 템플릿 엔진, view code(html)과 data logic code(db connection)를 분리해주는 기능을 수행
    - 레이아웃 템플릿 엔진과 텍스트 템플릿 엔진
        
        레이아웃 템플릿 엔진은 중복되는 include 코드를 사용하지 않고 지정된 페이지 레이아웃에 따라 페이지 타일을 조합하여 완전한 페이지를 만듦
        
        텍스트 템플릿 엔진은 템플릿 양식에 적절한 특정 데이터를 넣어 결과 문서를 출력
        
        주로 두 템플릿 엔진을 섞어서 많이 사용
        
    - 서버 사이드 템플릿 엔진과 클라이언트 사이드 템플릿 엔진
        
        서버 사이드 템플릿 엔진은 서버, DB, API에서 가져온 데이터를 미리 정의된 Template에 넣어 html을 그려 클라이언트에게 전달해주는 역할을 수행
        
        클라이언트 사이드 템플릿 엔진이란 html 형태로 코드를 작성할 수 있고 동적으로 DOM을 그리게 해주는 역할을 수행
        

### mvc란?

애플리케이션을 세가지 주요 논리적 구성 요소인 model, view, controller로 구분하틑 아키텍처/디자인 패턴

대규모 어플리케이션을 모두 고유한 목적을 가진 특정 세션으로 분할하여 사용자가 간편히 제어할 수 있도록 프로젝트를 구성하기 위해 사용함

### mvc의 특징

- business 로직, UI 로직, input 로직을 확실히 분리할 수 있음
- 웹 어플리케이션 아키텍쳐를 쉽게 디자인할 수 있도록 HTML과 URL에 대한 완벽한 제어가 가능
- TDD를 지원

### mvc의 구성 요소

- Controller: 뷰와 모델이 상호작용할 수 있도록 도와주는 구성 요소
    
    컨트롤러는 모델에게 데이터 로직, 비즈니스 로직, 입력 요청 등을 처리하도록 명령을 하고 뷰를 통해 최종 출력을 랜더링
    
    - 사용자의 입력을 받고 해석
    - 모델을 사용자의 요청에 따라 업데이터
    - 올바를 뷰를 선택하고 제공
- Model: 사용자가 사용하는 모든 데이터 관련 로직에 해당
    
    데이터베이스와 상호작용하며 컨트롤러의 요청에 의해 데이터를 제공하는 역할
    
    - 데이터 관리: CRUD(Create, Read, Update, Delete) operation
    - 뷰와 컨트롤러에게 상태 변화를 공지
- View: 어플리케이션의 모든 UI 로직을 담당
    
    유저 인터페이스를 사용자에게 제공하고 모델 컴포넌트로부터 제공받은 데이터를 통해 생성됨 (이 때 데이터는 모델에서 컨트롤러를 거쳐 전달)
    
    - 유저에게 데이터를 특정 형식으로 랜더링해 제공
    - 유저 인터페이스의 요소를 제공
    - 모델이 변화할 경우 출력을 업데이트

### mvc의 장점

- Design principles
    1. Divide and conquer: 모든 컴포넌트들은 독립적으로 디자인
    2. Increase conhension: 뷰와 컨트롤러가 하나의 UI 레이어에 존재에 레이어 응집력을 강화
    3. Reduce coupling: 세 개의 구성 요소들간의 소통이 최소화되고 간단
    4. Increase reuse
    5. Design fot flexibility: UI의 변환이 간단해 짐
- 각각의 컴포넌트들은 따로 테스트가 가능
- Search Engine Optimization (SEO)이 간편

### mvc의 단점

- 작은 어플리케이션엔 부적함
- 뷰에서의 데이터 접근이 비효율적
- 뷰와 모델 사이의 의존성이 높음

### MVC1

브라우저가 JSP page에 대한 요청을 보내고 JSP가 DB로부터 필요한 데이터를 받아 Model 객체(Java Bean)를 JSP 페이지(view)에 담아 응답으로 보내는 패턴

- JSP가 뷰와 컨트롤러의 역할을 모두 수행
- Bean은 모델의 역할을 수행
- 단점
    - 모든 페이지가 다음 페이지를 결정하는 로직을 담고 있어 jsp 페이지 이름이 다른 페이지에 의해 변경되거나 하는 상황이 발생하면 maintenance problem 발생
    - 확장이 어려움

## 2. (REST) URL 설계 규칙을 알고 계신가요?

Rest(Representational State Transfer): 웹을 이용할 때 제약 조건들을 정의하는 소프트웨어 아키텍처 스타일

RESTful API: HTTP URL을 통해서 자원을 명시하고 HTTP Method를 통해 해단 자원에 대한 CRUD를 적용하는 것

- REST 구성요소
    - 자원(Resource): HTTP URL
    - 자원에 대한 행위: HTTP Method
    - 자원에 대한 표현: Representation
- REST API 설계 규칙
    - URI에 후행 슬래시는 포함되지 않아야 함
    - 슬래시는 계층관계를 나타내는데 사용
    - URI의 가독성을 위해 하이픈을 사용
    - URI에서 언더바는 사용 금지
    - 소문자 선호
    - 파일 확장자는 포함되지 않아야 함
    - 마지막 부분은 복수형을 사용
    - 리소스에 대한 행위를 HTTP Method로 표현 (URI에 HTTP Method 포함 불가능, resource는 동사가 포함되서는 안되고 명사를 사용)

## 3. `Exception`과 `RuntimeException`의 차이에 대해 알고 계신가요?

- [[Spring의 Exception Handling Process](https://mangkyu.tistory.com/204)](https://mangkyu.tistory.com/204)에 대해 알고 계신가요?
    - 기본적인 예외 처리 방식
        
        Spring은 만들어질 때부터 에러 처리를 위한 BasicErrorController를 구현해두고 예외가 발생하면 기본적으로 /error로 에러 요청을 다시 전달하도록 WAS 설정
        
        - BasicErrorController: accept 헤더에 따라 에러 페이지를 반환하거나 에러 메세지를 반환
            
            기본 에러 경로는 /error → properties에서 servlet.error.path로 변경 가능
            
        - properties를 통해 에러 응답을 조정 가능
    - HandlerExceptionResolver
        
        에러 처리라는 공통 관심사를 메인 로직으로부터 분리하는 방법
        
        예외 처지 전략을 추상화한 인터페이스
        
        - 발생한 Exception을 catch하고 HTTP 상태나 응답 메시지 등을 설정
            
            WAS는 해당 요청이 정상적인 응답인 것으로 인식해 복잡한 에러 전달 과정이 반복되지 않음
            
        - HandlerExceptionResolver 구현체들을 빈으로 등록해 관리하고 적용 가능한 구현체를 찾아 예외 처리
        - ExceptionResolver를 동작시키는 도구
            
            ResponseStatus(에러 상태 변경), ResponseStatusException, ExceptionHandler, ControllerAdvice, RestControllerAdvice
            
- `ControllerAdvice`를 사용해보셨나요?
    
    특정 controller가 아닌 어플리케이션 전체의 익셉션을 처리하고 싶은 경우 사용, @RequestMapping이 달린 메서드 실행 중 발생한 예외를 처리
    
    @ControllerAdvice 어노테이션을 가진 클래스는 controller-advice로 내부의 @Component 어노테이션을 통해 스프링 빈으로 등록
    
    - 장점
        - 모든 컨트롤러의 예외를 하나의 클래스로 처리 가능
        - 일관성이 있는 에러 응답 가능

### RuntimeException

RuntimeException 클래스와 RuntimeException을 상속받은 클래스

프로그램 실행 중에 발생하는 익셉션

보통 프로그래밍 에러에 의해 발생하고 예측 불가능

NullPointerException, ClassCastException, IndexOutOfBoundsException, ArithmeticException 등이 해당

컴파일러에 의해 확인되지 않기 때문에 코드에서 익셉션을 올바르게 관리 해주어야 함

### Java Exception

 Exception 클래스와 Exception 클래스들을 상속 받은 나머지 클래스

Exception은 예측되지 않은 상태에 의해 프로그램 실행 도중에 발생하는 에러의 종류

exception handling이라는 프로세스에 의해 관리됨

- 익셉션이 발생했을 경우
    - 프로그램의 실행이 interrupte 되고, 프로그램 실행 제어 권한이 exception handler로 이동
    - exception handler는 익셉션을 handling하고 사용자에게 올바른 반응을 제공해야 하는 책임이 있음
        
        response에는 프로그램 종류, 에러 메시지 등이 있음
        

### Exception과 RuntimeException의 차이

컴파일시의 예외 처리 체크 유무

컴파일러는 RuntimeException을 제외한 모든 Exception 클래스들을 컴파일시 예외처리 여부를 반드시 확인하지만 RuntimeException 종류의 클래스가 발생했을 경우엔 확인을 하지 않아 컴파일이 가능

RuntimException은 주로 코드 상의 문제에서 발생하지만 Exception은 주로 외부 요인에 의해 발생 

## 4. HTTP Status Code를 잘 활용하고 계신가요?

- 애용하는 Status Code가 있나요?
- 본인에게는 익숙하지만, 누군가에겐 조금 생소할 만한 Code가 있다면 공유해주시겠어요? (like `429` [[Too Many Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429))
- [[여러분은 몇 번 코드가 옳다고 생각하시나요?](https://techblog.yogiyo.co.kr/%EB%8D%B0%EC%9D%B4%ED%84%B0%EA%B0%80-%EC%97%86%EC%9D%84-%EB%95%8C-200%EC%9D%B8%EA%B0%80-404%EC%9D%B8%EA%B0%80-f1c8c39ca9df)](https://techblog.yogiyo.co.kr/%EB%8D%B0%EC%9D%B4%ED%84%B0%EA%B0%80-%EC%97%86%EC%9D%84-%EB%95%8C-200%EC%9D%B8%EA%B0%80-404%EC%9D%B8%EA%B0%80-f1c8c39ca9df) (이건 정말 콜로세움 열립니다 ㅋㅋ)

### HTTP Status Code

- 1xx: Informational
- 2xx: Success
- 3xx: Redirection
- 4xx: Client Error
- 5xx: Server Error

## 5. [[Content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation)에 대해 알고 계시나요?

user agent가 어떤 URI에 있는 리소스를 요청할 때, 해당 리소스가 사용자에게 적절한 형태로 받도록 정하게 해주는 방법

### 서버 주도적 협상(Server-driven contect negotiation)

브라우저가 유저가 선호할만한 형태를 서술하는 여러개의 HTTP 헤더를 URL에 포함시켜 전송하면 내부 알고리즘이 최적의 컨텐츠를 사용자에게 전달

만약에 적절한 자원을 제공할 수 없다면 406(Not Acceptable)이나 415(Unsupported Media Type)로 응답하고 헤더를 서버 주도적 협상이 지원하는 헤더로 변경

- 단점: 서버는 브라우저에 대한 지식이 많지 않기 때문에 서버의 선택이 꽤 추상적

### 에이전트 주도적 협상(Agent-driven negotiation)

서버가 모호한 요청에 마주했을 때 서버는 대체 가능한 링크를 제공하는 방법

구체적으로 서버는 user agent가 바고자 하는 리소스의 형태와 메타데이터에는 어떤 것이 있는지 user agent에게 정보를 제공

user agent는 여기서 알맞은 형태를 고르고, 해당 요청을 받기 위한 특별한 URL로 서버에 재요청

## 6. TLS/SSL에 대해 알고 계신가요?

- HTTPs는 Request/Response의 어떤 부분을 암호화하나요? 전체? Body? 어쩌면 Start Line을 제외한 나머지?
    
    Reqeust와 Response 전체를 암호화 → Start Line, Headers, Body를 포함한 모든 부분 
    
    도메인 이름은 TLS 연결을 설정하기 전에 서버를 찾아야하기 때문에 암호화하지 않음
    

### TLS(Transport Layer Security)

인터넷 상의 커뮤니케이션을 위한 개인 정보와 데이터 보안을 용이하게 하기 위해 설계되어 널리 채택된 보안 프로토콜

암호화, 인증, 무결성이라는 세 가지 요소를 달성

- 비대칭 공개 키 인프라로 사용
    - 개인 키 - 웹 사잍트 소유자가 관리, 비공개로 유지되며 웹 서버에 있어 공개 키로 암호화된 정보를 해독하는 데에 사용
    - 공개 키 - 안전한 방식으로 서버와 상호 작용하고자하는 모든 사람이 사용 가능, 공개 키로 암호화된 정보는 개인 키로만 해독 가능

### SSL(Secure Sockets Layer)

웹 사이트와 브라우저 사이에 전송되는 데이터를 암호화하여 인터넷 연결을 보호하기 위한 표준 기술

TLS와 SSL은 같은 의미, TLS가 SSL의 후속 버전일 뿐

### HTTPS

웹 브라우저와 웹 사이트 간에 데이터를 전송하는데 사용되는 기본 프로토콜인 HTTP의 보안 버전

‘https’를 통한 인터넷 접속은 우리 컴퓨터가 해당 사이트의 서버와 암호화 통신을 하고 있다는 것을 의미하고 SSL을 이용한 것

- 대칭키 암호화 기법을 사용
- HTTP 자체에는 암호화하는 기능이 없기 때문에 HTTPS가 HTTP 메시지를 TLS 프로토콜을 통해 암호화하여 보호

## 7. OpenAPI spec에 대해 들어보셨나요?

- (최신) Swagger는 GET request의 payload가 request body에 담기는 경우를 허용하지 않는답니다...
- [[JSON reponse spec에 대한 논의](https://stackoverflow.com/questions/12806386/is-there-any-standard-for-json-api-response-format)](https://stackoverflow.com/questions/12806386/is-there-any-standard-for-json-api-response-format)도 재답니다.

**Comment by SIKU-KR** [comment]
### 1. 전통적인 MVC 구조를 알고 계신가요?
Model-View-Controller의 구조로서, 
Model : 데이터에 접근하는 역할을 담당하는 객체
View : 사용자와 상호작용하는 역할을 담당하는 객체
Controller : View와 Controller의 사이에서 흐름을 통제하며 비즈니스 로직을 포함하는 객체

### 2. (REST) URL 설계 규칙을 알고 계신가요?
서버에서 접근하는 리소스를 명사로 표기하는것이 규칙
URI에서는 리소스와 행위를 분리 시켜야한다. 행위를 메서드로 구분함
- GET: 리소스 조회
- POST: 요청 데이터 처리, 주로 등록
- PUT: 리소스를 대체, 해당 리소스가 없으면 생성
- PATCH: 리소스 부분 변경
- DELETE: 리소스 삭제

### 3. Exception과 RuntimeException의 차이에 대해 알고 계신가요?
- Exception
  - 컴파일 중에 발생한다.
  - 반드시 예외처리가 되어야한다.
  - 트랜잭션 롤백이 되지않음.
  - 외부의 영향으로 인해 발생할 수 있음(사용자동작, 운영체제 등)
- Runtime-Exception
  - 런타임에 발생한다.
  - 예외처리를 하지 않아도 된다.
  - 트랜잭션 롤백이 진행된다.
  - 프로그래머의 실수로 인해 발생한다.

#### Spring의 Exception Handling Process에 대해 알고 계신가요?
Spring은 예외 발생 시 HandlerExceptionResolver를 이용해 예외를 처리합니다. 
기본적으로 제공되는 Resolver
- ExceptionHandlerExceptionResolver: @ExceptionHandler 메서드를 찾고 처리합니다.
- ResponseStatusExceptionResolver: 예외 클래스에 @ResponseStatus가 붙어 있는 경우 이를 처리하여 적절한 HTTP 응답 상태 코드를 반환합니다.
- DefaultHandlerExceptionResolver: 스프링에서 발생하는 기본적인 예외를 처리하며 404, 405, 500 등의 상태 코드를 반환합니다.

#### ControllerAdvice를 사용해보셨나요?
전역적으로 예외를 처리하고자 할 때는 @ControllerAdvice를 사용합니다. 특정 예외나 컨트롤러를 지정하여 전역적인 예외 처리를 설정할 수 있습니다.

```java
@ControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(MyCustomException.class)
    public ResponseEntity<String> handleMyException(MyCustomException ex) {
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Global custom error message");
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<String> handleException(Exception ex) {
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("An error occurred");
    }
}
```

### 4. Content negotiation에 대해 알고 계시나요?

클라이언트가 선호하는 표현 요청
- Accept : 클라이언트가 선호하는 미디어 타입 전달
- Accept-Charset: 클라이언트가 선호하는 문자 인코딩
- Accept-Encoding: 클라이언트가 선호하는 압축 인코딩
- Accept-Language: 클라이언트가 선호하는 자연언어
   
### 5. TLS/SSL에 대해 알고 계신가요?
TLS(Transport Layer Security)와 SSL(Secure Sockets Layer)은 네트워크 통신에서 **데이터를 암호화**하여 **보안을 강화**하는 프로토콜입니다. TLS는 SSL의 업그레이드된 버전으로, SSL에서 발견된 여러 취약점을 개선하고 보안을 강화한 프로토콜입니다.

SSL은 웹 브라우저와 웹 서버 간에 데이터를 안전하게 전송하기 위해 처음 개발된 **암호화 프로토콜**입니다. SSL은 **넷스케이프(Netscape)**에서 1990년대 초반에 개발되었으며, 이후 SSL 1.0, 2.0, 3.0이 연이어 나왔습니다. 

TLS는 SSL의 뒤를 이은 보안 프로토콜로, SSL의 **보안 문제를 해결**하고 더 안전한 암호화 방법을 도입했습니다. 1999년에 **TLS 1.0**이 발표되었으며, 이후 TLS 1.1, TLS 1.2, TLS 1.3이 차례로 발표되면서 보안이 크게 강화되었습니다.

#### HTTPs는 Request/Response의 어떤 부분을 암호화하나요? 전체? Body? 어쩌면 Start Line을 제외한 나머지?

도메인과 포트번호를 제외한 모든 부분이 암호화된다.
- Request Line (Start Line): HTTP 메서드(GET, POST 등)와 요청 경로(/path), 그리고 HTTP 버전은 TLS 연결 이후 암호화됩니다.
- Headers: HTTP Headers도 모두 암호화됩니다. 예를 들어, Authorization 헤더나 Cookie 헤더에 민감한 정보가 있을 경우 암호화되기 때문에 안전하게 전송됩니다.
- Body: POST나 PUT 요청의 본문(body)도 암호화됩니다. 예를 들어, 사용자가 폼을 제출하거나 파일을 업로드할 때의 데이터가 포함됩니다.
